---
title: "Identification of influenza biomarkers from transcriptomics data"
author: "Diogo M. Camacho"
date: "`r format(Sys.Date(), '%Y-%m-%d')`"
output:
  html_document:
    df_print: paged
subtitle: 'Dataset: 2018-05-29'
abstract: Here we will be exploring data set for the flu project (a collaboration
  with the Ingber lab). The data corresponds to an analyses of transcriptomics data
  collected from one healthy donor lung cells, with cells being cultured in transwells
  (no endothelial cells.) Additionally, we will be looking at publicly available data
  for identification of potential host biomarkers for influenza infection. RNA-seq
  data was collected by the DFCI transcriptomics core.
---

```{r libsfun, echo = FALSE, include=FALSE}
# libraries
library(tidyverse)
library(readr)
library(readxl)
library(DRUID)
library(grove)
library(DESeq2)
library(org.Hs.eg.db)
library(rpegeos)
library(diffnet)
library(plotly)
library(igraph)
library(ggraph)

# functions
source("R/stouffer_specificity.R")
#source("R/limma_dge.R")
```


# Load data
```{r load_data}
# data from project ----
# count data 
count_data <- read_csv(here::here("data/rna_seq/2018-05-29_STAR_Gene_Counts.csv"))
gene_ids <- count_data$Gene_ID
count_data <- count_data %>% dplyr::select(., -Gene_ID)

# metadata
metadata <- readxl::read_xlsx("data/rna_seq/2018-05-29_RNAseq_Sample IDs.xlsx")
t1 <- gsub("5/2", "May2", gsub("5/1", "May1", gsub("-", "_", metadata$`Tube ID`)))

metadata <- metadata %>% 
  dplyr::mutate(., sample_name = t1)

metadata <- metadata[match(colnames(count_data), metadata$sample_name), ]

group <- character(length = nrow(metadata))
group[metadata$Sample == "Control"] <- "control"
group[metadata$Sample == "Virus Treated"] <- "virus"
group[metadata$Sample == "Poly IC Treated"] <- "poly_ic"

metadata <- metadata %>% 
  dplyr::mutate(., group = group)

sample_data <- tibble(sample_name = metadata$sample_name,
                      time_point = as.numeric(gsub("h", "", metadata$`Time point`)),
                      group = metadata$group)

sample_data <- sample_data %>% 
  dplyr::mutate(., group_time = paste(group, time_point, sep = "_"))
```


# Processing data 
Here we will perform a set of data clean up steps. These will include:

 - removing probes that show zero counts across all samples
 - remove genes with multiple EntrezIDs (I consider these to be poor mappings -- non-specific)
 - remove genes with NA entrez ids (can't map these to pathways/processes)

## Remove zero counts
```{r zero_counts}
# remove probes with zero counts throughout
nix <- which(rowSums(count_data) == 0)
if(length(nix) != 0) {
  gene_ids <- gene_ids[-nix]
  count_data <- count_data[-nix, ]
}
```

## Multiple EntrezIDs
```{r mult_ent}
# remove gene ids with multiple entrez mappings
# get annotations
gkey <- AnnotationDbi::mapIds(x = org.Hs.eg.db, 
                                keys = gene_ids, 
                                column = "ENTREZID",
                                keytype = "SYMBOL")

gene_df <- tibble::tibble(gene_symbol = gene_ids,
                          entrez_id = gkey)
```

## Genes with NA EntrezID
```{r na_genes}
# remove genes with NA entrez id
nix <- which(is.na(gene_df$entrez_id))
if(length(nix) != 0) {
  gene_df <- gene_df[-nix, ]
  count_data <- count_data[-nix, ]
}
```

## Frequency based filtering
Finally, we will filter genes based on their frequency in the data set. What this entails is performing the following filtering steps:

 - Calculate frequency of genes across samples (ie: in how many samples is the gene measured?). We set this to 75% of the samples.
 - Establish a minimum set of counts for a gene to be considered (set at 5 counts in any given sample)
 
By combining these 2 conditions, we are only including genes that at least 5 counts in at least 75% of the samples (which corresponds to at least `r round(0.75 * dim(count_data)[2])` samples.)

```{r freq_filt}
# frequency across samples
# minimum 5 read counts, at least in 75% of the samples
x1 <- apply(count_data, 1, function(y) length(which(y >= 5)))
x1 <- which(x1 >= round(0.75 * ncol(count_data)))

count_data <- count_data[x1, ]
gene_df <- gene_df[x1, ]
# gkey <- AnnotationDbi::select(x = org.Hs.eg.db, 
#                               keys = gene_ids, 
#                               columns = c("ENTREZID", "GENENAME"),
#                               keytype = "SYMBOL")
```

# Differential expression
With the data processed as described in the steps above, we can now proceed with standard analyses such as calculating the differential expression for the genes given the conditions we have. For the differential expression we will use the `DESeq2` package, as this is the state-of-the-art tool for differential analyses of RNA-seq data. 

```{r deseq2}
dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = sample_data,
                              design = ~ group_time)

dds <- DESeq(object = dds)
```

This establishes the `DESeq2` object that we can now query with different contrasts. For the purpose here we will perform the following comparisons:

 - Virus infected @ 18h versus Control @ 18h
 - Poly-IC exposure @ 18h versus Control @ 18h
 
Poly-IC is a TLR3 agonist and is used as a control as a standard immunostimulant. We will use this to our advantage as a differentiator to allow us for the characterization of the viral-specific response. 

```{r dge_comparisons}
contrast_v18 <- c("group_time", "virus_18", "control_18")
contrast_ic18 <- c("group_time", "poly_ic_18", "control_18")

# viral response
res_v18 <- results(dds,
        contrast = contrast_v18,
        pAdjustMethod = "fdr",
        cooksCutoff = FALSE)

res_v18 <- res_v18 %>% 
  as_tibble() %>%
  tibble::add_column(., gene = gene_df$gene_symbol) %>% 
  tibble::add_column(., entrez_id = gene_df$entrez_id) %>% 
  tibble::add_column(., sig = 0) %>%
  dplyr::mutate(., sig = replace(sig, list = which(padj < 0.01 & abs(log2FoldChange) > 1), 1))

# polyic response
res_ic18 <- results(dds,
                   contrast = contrast_ic18,
                   pAdjustMethod = "fdr",
                   cooksCutoff = FALSE)

res_ic18 <- res_ic18 %>%
  as_tibble() %>%
  tibble::add_column(., gene = gene_df$gene_symbol) %>% 
  tibble::add_column(., entrez_id = gene_df$entrez_id)
```

We can look at what these comparisons look like using an MA-plot or a volcano plot:

```{r volcano_plots, plotly = TRUE}
p1 <- res_v18 %>% 
  ggplot() + 
  geom_point(aes(y = log2FoldChange, x = log10(baseMean), color = as.factor(sig), gene = gene, entrez = entrez_id, fold_change = log2FoldChange, base_mean = log10(baseMean)), alpha = 0.5, size = 3) + 
  ylim(c(-max(abs(res_v18$log2FoldChange)), max(abs(res_v18$log2FoldChange)))) +
  # scale_color_viridis_d() +
  scale_color_manual(breaks = c(0, 1), name = NULL, values = c("black", "red"), labels = c("Not significant", "|F| > 1, p < 0.01")) +
  labs(y = "log(fold change)", x = "Mean normalized counts", title = "MA plot: Viral response") + 
  theme_bw() + 
  theme(axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black"),
        axis.ticks.length = unit(0.2, "cm"),
        legend.position = "top")
ggsave(plot = p1, filename = here::here(paste0("res/dataset_1/", format(Sys.Date(), "%Y-%m-%d"), "_MAplot.tiff")), device = "tiff", width = 11, height = 8, dpi = 600)

p2 <- res_v18 %>% 
  ggplot() + 
  geom_point(aes(x = log2FoldChange, y = -log10(padj), color = as.factor(sig), gene = gene, entrez = entrez_id, fold_change = log2FoldChange, padj = -log10(padj)), alpha = 0.5, size = 3) + 
  ylim(c(0, max(-log10(res_v18$padj)))) +
  xlim(c(-max(abs(res_v18$log2FoldChange)), max(abs(res_v18$log2FoldChange)))) +
  # scale_color_viridis_d() +
  scale_color_manual(breaks = c(0, 1), name = NULL, values = c("black", "red"), labels = c("Not significant", "|F| > 1, p < 0.01")) +
  labs(y = "log(fold change)", x = "Mean normalized counts", title = "Volcano plot: Viral response") + 
  theme_bw() + 
  theme(axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black"),
        axis.ticks.length = unit(0.2, "cm"),
        legend.position = "top")
ggsave(plot = p2, 
       filename = here::here(paste0("res/dataset_1/", format(Sys.Date(), "%Y-%m-%d"), "_volcano-plot.tiff")), 
       device = "tiff", 
       width = 11, 
       height = 8, 
       dpi = 600)

p1

p2

# # now interactive plots
# plotly::ggplotly(p1, tooltip = c("gene", "entrez_id", "fold_change", "base_mean"))
# 
# plotly::ggplotly(p2, tooltip = c("gene", "entrez_id", "fold_change", "padj"))
```


# Virus-specific transcriptional changes

Now we can look at which genes are differentially expressed specifically in the context of a viral infection. Here is where we will use the poly IC to our advantage, as we will only consider relevant those genes that are differentially expressed in the viral infection case that are not affected in the poly IC treatment:

```{r viral_specificity}
# viral specific genes
fold_thr <- 1
pval <- 0.05

# virus_sig <- with(res_v18, which(padj < 0.05 & log2FoldChange > 1))
virus_sig <- res_v18 %>% 
  dplyr::filter(., abs(log2FoldChange) > fold_thr, padj < pval) 

ic_sig <- res_ic18 %>% 
  dplyr::filter(., abs(log2FoldChange) > fold_thr, padj < pval) 


# virus_specific <- setdiff(virus_sig, ic_sig)
virus_specific <- virus_sig %>% 
  tibble::add_column(., ic_gene = 0) %>% 
  dplyr::mutate(., ic_gene = replace(ic_gene, which(gene %in% ic_sig$gene), 1)) %>% 
  dplyr::filter(., ic_gene == 0)
```

This gives us a set of `r nrow(virus_specific)` genes that are differentially expressed in the viral treatment that are specific to that response:

```{r}
virus_specific
```

We can also visualize these results in an MA plot:

```{r plot_virus_specific_ma}
p3 <- res_v18 %>% 
  dplyr::mutate(sig, sig = replace(sig, which(entrez_id %in% virus_specific$entrez_id), 2)) %>%
  ggplot() + 
  geom_point(aes(y = log2FoldChange, x = log10(baseMean), color = as.factor(sig), gene = gene, entrez = entrez_id, fold_change = log2FoldChange, base_mean = log10(baseMean)), alpha = 0.5, size = 3) + 
  ylim(c(-max(abs(res_v18$log2FoldChange)), max(abs(res_v18$log2FoldChange)))) +
  scale_color_viridis_d(breaks = c(0, 1, 2), name = NULL, labels = c("Not significant", "Significant, general", "Significant, virus specific")) +
  # scale_color_manual(breaks = c(0, 1, 2), name = NULL, values = c("black", "red", "tan"), labels = c("Not significant", "|F| > 1, p < 0.01", "Significant, virus specific")) +
  labs(y = "log(fold change)", x = "Mean normalized counts") + 
  theme_bw() + 
  theme(axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black"),
        axis.ticks.length = unit(0.2, "cm"),
        legend.position = "top")
ggsave(plot = p3, 
       filename = here::here(paste0("res/dataset_1/", Sys.Date(),"_virus-specific_MAplot.tiff")),
       device = "tiff", 
       width = 11, 
       height = 8, 
       dpi = 600)

p3


```



# Identifying compounds to enhance host response to viral infection

With the identified set of host genes that are specifically differentially expressed under viral treatment, we can use DRUID to identify existing compounds that could be used to revert this phenotypic response. 

```{r druid, cache = FALSE}
# build gene set for DRUID ----
druid_genes <- cbind(virus_specific$log2FoldChange,
                     virus_specific$padj)

# druid parameters
res <- vector(mode = "list", length = 5)
for (i in seq(1, 5)) {
      res[[i]] <- DRUID::run_druid(dge_matrix = druid_genes, 
                            druid_direction = "neg", 
                            fold_thr = 0, 
                            pvalue_thr = 0.05, 
                            entrez = virus_specific$entrez_id, 
                            num_random = 10000, 
                            selection = i)
}
flu_druid <- dplyr::bind_rows(res)
rm(res)

dev <- "tiff"
p4 <- flu_druid %>%
  dplyr::filter(., druid_score > 5, number_matches > 1) %>%
  # dplyr::slice(1:1000) %>%
  ggplot() + 
  geom_point(aes(x = druid_score, y = forcats::fct_reorder(drug_name, druid_score), color = data_source), alpha = 0.5, size = 4) +
  # geom_point(aes(x = druid_score, y = drug_name, color = druid_score), alpha = 0.5, size = 4) +
  # scale_color_viridis_c() +
  scale_color_viridis_d() +
  # facet_grid(. ~ data_source, scales = "free") +
  labs(x = "DRUID score", y = "Drug") +
  theme_bw() +
  theme(axis.title = element_text(size = 24, color = "black"),
        strip.text.x = element_text(size = 12, color = "black"),
        axis.text = element_text(size = 12, color = "black"),
        axis.ticks.length = unit(0.2, "cm"))
ggsave(plot = p4,
       filename = paste0("res/dataset_1/", Sys.Date(),"_DRUID-flu_transwells.", dev),
       device = dev,
       height = 8,
       width = 11,
       dpi = 600)

p4

flu_druid


```



# Tissue specificity

Additionally, we can use publicly available data to assess which genes are preferentially expressed in a given tissue. For this particular problem, we are interested in identifying the genes that are mostly expressed in lung. For that, we will use the human expression atlas data that is housed at [BioGPS](http://www.biogps.org). We can also use data from [GTex](https://www.gtexportal.org/home/), though this is not implemented yet (will include this in future work.)

The tissue specificity is calculated using an adaptation of Stouffer's method. Briefly, we normalize the expression data along the gene **and** the sample spaces, independently, and we combine the normalized scores into a single representation given by:

$ Z_{global} = \frac{Z_{genes} + Z_{samples}}{\sqrt(2)} $

where $Z_{global}$ is the tissue specificity score, $Z_{genes}$ corresponds to the normalized data along the gene space (normalizing along the columns), and $Z_{samples}$ is the normalized data along the sample space (normalizing along the rows). 

First, we will compute the average expression of any given gene for any given tissue. 

```{r average_tissue_expression}
# biogps data ----
load(here::here("data/2019-02-13_biogps.RData"))


# average of tissues ----
utis <- unique(biogps_data$tissues)
avg_E <- matrix(0, nrow = nrow(biogps_data$expression), ncol = length(utis))
for (i in seq(1, length(utis))) {
  u1 <- which(biogps_data$tissues == utis[i])
  if (length(u1) > 1) {
    avg_E[, i] <- rowMeans(biogps_data$expression[, u1])
  } else {
    avg_E[, i] <- biogps_data$expression[, u1]
  }
}

```


With these data we can now computer the tissue specificity:

```{r stouffer_specificity}
# lung specific genes ----
stouf_thr <- 3
ts_stouffer <- stouffer_specificity(avg_E)
```

As we can see, the dimensions of the (average) expression matrix (ncols = `r ncol(avg_E)`; nrows = `r nrow(avg_E)`) and the tissue specificity matrix (ncols = `r ncol(ts_stouffer)`; nrows = `r nrow(ts_stouffer)`) are exactly the same, which is what we wanted. Let's now look at what this transformation does, by sampling a set of genes for illustration purposes only.



```{r specificity_visualization}
thr <- 3

A <- ts_stouffer
A[which(A < thr)] <- 0

a1 <- apply(A, 1, function(x) length(which(x != 0)))

a2 <- which(a1 < 3 & a1 != 0)
a3 <- sample(x = a2, size = 25, replace = FALSE)

example_data <- tibble::tibble(gene = rep(biogps_data$genes$SYMBOL[a3], ncol(A)),
                               tissue = as.vector(sapply(utis, rep, length(a3))),
                               specificity = as.vector(A[a3, ]))

example_data %>% 
  ggplot(aes(x = tissue, y = gene, fill = specificity)) +
  geom_tile() + 
  scale_fill_viridis_c() + 
  labs(x = "Gene", y = "Tissue") + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, size = 12, color = "black", hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 12, color = "black"), 
        axis.title = element_text(size = 12, color = "black"),
        panel.grid = element_blank())

```

These results can be cross-referenced with the data on the [Human Protein Atlas](https://www.proteinatlas.org), as another source of tissue specificity for human genes. With these data, we can now look at genes that have specific (or preferential) expression in lung tissue when compared to other tissues. 

## Lung specific genes
Using the transformation above, we will look for genes that are primarily expressed in the lung. 

```{r lung_specific, cache=FALSE}
B <- ts_stouffer
B[which(B < thr)] <- 0
B[B != 0] <- 1

num_tissues <- rowSums(B)

spec_genes <- which(num_tissues <= 2 & num_tissues != 0) # <-- gives at most 2 tissues with high specificity for a given gene

# now make a data frame for these:
specificity_df <- tibble::tibble(gene_symbol = rep(biogps_data$genes$SYMBOL[spec_genes], length(utis)),
                                 entrez_id = rep(biogps_data$genes$ENTREZID[spec_genes], length(utis)),
                                 tissue_name = as.vector(sapply(utis, rep, length(spec_genes))),
                                 presence_call = as.vector(B[spec_genes, ]))

specificity_df <- specificity_df %>%
  dplyr::filter(., presence_call != 0)

lung_genes <- specificity_df %>%
  dplyr::filter(., tissue_name == "lung") %>%
  dplyr::select(., -presence_call)

lung_genes
```

## Secreted biomarkers
Additionally, we can use the annotations of the different proteins to identify those that are: 

  1. Secreted proteins that respond to influenza infection
  2. Lung specific genes that have differential response to influenza infection
  
The rationale behind this approach is for the identification of biomarkers that can be used for diagnostics of infection states. 

```{r secreted_biomarkers}
# signal peptides ----
sig_pep <- read_xlsx(here::here("data/secretome/2019-03-07_signal_peptides_human.xlsx"))

sig_peps_names <- AnnotationDbi::select(x = org.Hs.eg.db, 
                                        keys = as.character(sig_pep$`Accession Number`),
                                        keytype = "UNIPROT", columns = c("ENTREZID", "SYMBOL")) %>% 
  as_tibble() %>% 
  dplyr::filter(., !is.na(ENTREZID))


# differentially abundant signal peptides in our data
# virus_sig_peps <- intersect(sig_peps_names$SYMBOL, gene_ids[virus_specific])

# secreted proteins ----
# based on HPA predicted data
hpa_pred <- read_delim(here::here("data/secretome/2019-03-07_HPA_predicted_secreted_proteins.tsv"), delim = "\t")

hpa_pred <- tibble::tibble(gene = hpa_pred$Gene,
                           synonym = hpa_pred$`Gene synonym`,
                           description = tolower(hpa_pred$`Gene description`),
                           class = tolower(hpa_pred$`Protein class`), 
                           evidence = tolower(hpa_pred$Evidence),
                           antibody = hpa_pred$Antibody,
                           subcellular = hpa_pred$`Subcellular location`) %>%
  dplyr::filter(., grepl(pattern = "protein level", evidence))

# predicted secreted biomarkers ----
# virus_biom_sec <- union(virus_secp, virus_sig_peps)
virus_biomarkers <- virus_specific %>% 
  tibble::add_column(., signal_peptide = 0) %>% 
  tibble::add_column(., secreted = 0) %>%
  tibble::add_column(., lung_specific = 0) %>%
  dplyr::mutate(., signal_peptide = replace(signal_peptide, which(gene %in% sig_peps_names$SYMBOL), 1)) %>% 
  dplyr::mutate(., secreted = replace(secreted, which(gene %in% hpa_pred$gene), 1)) %>%
  dplyr::mutate(., lung_specific = replace(lung_specific, which(gene %in% lung_genes$gene_symbol), 1))

# plot ----
dev <- "tiff"
p1 <- virus_biomarkers %>% 
  dplyr::filter(., signal_peptide == 1 | secreted == 1) %>%
  ggplot() + 
  geom_point(aes(x = forcats::fct_reorder(gene, log2FoldChange, .desc = TRUE), y = log2FoldChange, color = log2FoldChange, size = -log10(padj)), alpha = 0.5) + 
  scale_color_viridis_c() + 
  labs(x = NULL, y = "log2(Fold change)", title = "Secreted proteins") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 24, color = "black"))
ggsave(plot = p1, 
       filename = paste0("res/dataset_1/", Sys.Date(),"_virus-specific-SECRETED-biomarkers.", dev),
       device = dev, 
       width = 11, 
       height = 8, 
       dpi = 600)

p2 <- virus_biomarkers %>% 
  # dplyr::filter(., signal_peptide == 1 | secreted == 1) %>%
  dplyr::filter(., lung_specific == 1) %>%
  ggplot() + 
  geom_point(aes(x = forcats::fct_reorder(gene, log2FoldChange, .desc = TRUE), y = log2FoldChange, color = log2FoldChange), size = 5, alpha = 0.5, stroke = 2, pct = 2) + 
  ylim(c(0, max(virus_biomarkers$log2FoldChange))) +
  scale_color_viridis_c() + 
  labs(x = NULL, y = "log2(Fold change)", title = "Lung specific") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 24, color = "black"))
ggsave(plot = p2, 
       filename = paste0("res/dataset_1/", Sys.Date(),"_virus-specific-LUNG-SPECIFIC-biomarkers.", dev),
       device = dev, 
       width = 11, 
       height = 8, 
       dpi = 600)

p1
p2


```


# Network representation
Finally, we will look at a representation of all these results in the context of a manually curated gene regulatory network for human genes (from [TTRUST](https://www.grnpedia.org/trrust/)). Given the genes in the virus specific group, their classification (secreted or not), and their class (regulator or not), this network can give us some sense as to what of the overarching biology at play from the host side.


```{r network_results}
# TTRUST DATA 
load(here::here("data/2018-04-30_ttrust_data.RData"))

a1 <- ttrust_data %>%
  dplyr::filter(., target %in% virus_specific$gene | tf %in% virus_specific$gene)

net <- a1 %>% 
  igraph::graph_from_data_frame()

v_name <- unique(c(a1$tf, a1$target))
v_color <- rep("gray", length(v_name))
v_shape <- rep("circle", length(v_name))
v_color[which(v_name %in% virus_specific$gene)] <- "red"
v_shape[which(v_name %in% a1$tf)] <- "square"

igraph::vertex_attr(net) <- list(name = v_name,
                                 color = v_color, 
                                 size = rep(5, length(igraph::V(net))),
                                 shape = v_shape)

p <- net %>%
  # ggraph(layout = "kk",
  #        maxiter = 10000,
  #        dim = 2,
  #        epsilon = 0.0001,
  #        kkconst = 0.1) +
  # ggraph(layout = "fr") +
  ggraph(layout = "nicely") +
  geom_edge_link(start_cap = circle(1.75, "mm"),
                 end_cap = circle(1.75, "mm"),
                 color = "black", 
                 show.legend = FALSE) +
  geom_node_point(#aes(fill = color), 
                  alpha = 1, 
                  # size = tmp3$node_size,
                  size = 5,
                  shape = igraph::vertex.attributes(net)$shape,
                  color = igraph::vertex.attributes(net)$color,
                  # color = "black",
                  # fill = igraph::vertex.attributes(net)$color,
                  # fill = "black", 
                  stroke = 1.25,
                  show.legend = FALSE) +
  geom_node_text(label = igraph::vertex.attributes(net)$name) +
  theme_minimal() + 
  guides(fill = guide_legend(override.aes = list(size = 12))) +
  theme(title = element_text(size = 32, color = "white"),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_rect(color = NULL, fill = NULL),
        plot.background = element_rect(fill = NULL),
        legend.text = element_text(size = 24, color = "black"),
        legend.position = "right") 


p



```


# Future work

 - Integration with Zaas et al data (2009)
 - Integration with COPD donor data
 - Exploration of network influences in defining infection response biology from host


<!--- Pathway enrichments 
```{r pathway_enr}

# FOR LATER
#pagerank of genes ----
# tmp1 <- igraph::page_rank(graph = net, directed = TRUE)

# genes we're interested in ----
# xx <- sort(tmp1$vector[match(virus_specific$gene, 
#                              names(tmp1$vector), 
#                              nomatch = FALSE)], 
#            decreasing = TRUE)
# 
# net2 <- igraph::as_data_frame(net)
# net2 %>% 
#   dplyr::filter(., from %in% names(xx) | to %in% names(xx))


# annotate genes ----
# yy <- AnnotationDbi::select(x = org.Hs.eg.db, keys = names(xx), columns = "ENTREZID", keytype = "SYMBOL")


# get expression data for those genes ----
# z1 <- match(yy$SYMBOL, gene_ids)
# z1 <- z1[!is.na(z1)]
# z2 <- cbind(gene_ids[z1], res_v18$log2FoldChange[z1], res_v18$padj[z1])
# z3 <- AnnotationDbi::select(x = org.Hs.eg.db, 
#                             keys = z2[, 1],
#                             keytype = "SYMBOL",
#                             columns = "ENTREZID")
# z4 <- cbind(z3$ENTREZID, z2[, c(2, 3)])
# z4 <- apply(z4, 2, as.numeric)

# enrich ----
# z5 <- rpegeos::enrich_geneset(z4)
# z5 <- oranges::oranges(query_entrez = z4[, 1], universe_entrez = as.vector(AnnotationDbi::mapIds(x = org.Hs.eg.db, keys = names(tmp1$vector), column = "ENTREZID", keytype = "SYMBOL")))
# 
# dev <- "tiff"
# z5 %>% 
#   dplyr::filter(., data_source == "go:bp") %>% 
#   dplyr::arrange(., desc(-log10(padj))) %>% 
#   dplyr::slice(1:25) %>%
#   ggplot() + 
#   geom_point(aes(x = -log10(padj), y = forcats::fct_reorder(name, -log10(padj)), color = -log10(padj), size = number_genes), alpha = 0.5) +
#   scale_color_viridis_c() + 
#   labs(x = "-log10(adjusted p-value)", y = "Pathway name") +
#   theme_bw() + 
#   theme(axis.title = element_text(size = 24, color = "black"),
#         axis.text = element_text(size = 12, color = "black"),
#         axis.ticks.length = unit(0.2, "cm"),
#         legend.position = "none")


# pathway enrichment ----
# using oranges (fisher exact test implementation)
# library(oranges)
# 
# chip_enr_ora <- oranges::oranges(query_entrez = virus_specific$entrez_id, universe_entrez = gkey$ENTREZID)
# dev <- "tiff"
# chip_enr_ora %>% 
#   dplyr::filter(., data_source == "go:bp") %>% 
#   dplyr::arrange(., desc(-log10(padj))) %>% 
#   dplyr::slice(1:25) %>%
#   ggplot() + 
#   geom_point(aes(x = -log10(padj), y = forcats::fct_reorder(name, -log10(padj)), color = -log10(padj), size = number_genes), alpha = 0.5) +
#   scale_color_viridis_c() + 
#   labs(x = "-log10(adjusted p-value)", y = "Pathway name") +
#   theme_bw() + 
#   theme(axis.title = element_text(size = 24, color = "black"),
#         axis.text = element_text(size = 12, color = "black"),
#         axis.ticks.length = unit(0.2, "cm"),
#         legend.position = "none")
# ggsave(filename = paste0("res/", Sys.Date(),"_chip-data-ORANGES-enrichment.", dev),
#        device = dev,
#        height = 8,
#        width = 11,
#        dpi = 600)
# 
# # 
# flu_enr_rpegeos <- rpegeos::enrich_geneset(gene_set = cbind(virus_specific$entrez_id, virus_specific$log2FoldChange, virus_specific$padj))
# dev <- "tiff"
# chip_enr %>%
#   dplyr::slice(1:20) %>%
#   ggplot() +
#   geom_point(aes(y = enrichment_score, x = forcats::fct_reorder(geneset, enrichment_score), size = number_genes), alpha = 0.5, color = "#ff6600") +
#   labs(x = "Gene set", y = "Enrichment score") +
#   coord_flip() + 
#   theme_bw() + 
#   theme(axis.title = element_text(size = 12, color = "black"),
#         axis.text = element_text(size = 12, color = "black"),
#         legend.position = "none")
# ggsave(filename = paste0("res/", Sys.Date(),"_chip-data-enrichment.", dev),
#        device = dev,
#        height = 8,
#        width = 11,
#        dpi = 600)
# 
# 
# flu_enr_cp <- clusterProfiler::enrichGO(gene = virus_specific$entrez_id, ont = "BP", pAdjustMethod = "fdr", universe = gkey$ENTREZID, OrgDb = org.Hs.eg.db)
```
--->
